File: port_Java/grun4py.java
// ******* GRUN (Grammar Unit Test) for Python *******

import org.antlr.v4.runtime.*;

import java.io.*;
import java.nio.charset.Charset;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

public class grun4py {
    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Error: Please provide an input file path");
            System.exit(1);
        }

        try {
            final Path path = Paths.get(args[0]);
            CharStream input = CharStreams.fromPath(path, Charset.forName("utf-8"));
            PythonLexer lexer = new PythonLexer(input);
            CommonTokenStream tokens = new CommonTokenStream(lexer);
            PythonParser parser = new PythonParser(tokens);

            tokens.fill(); // Test the lexer grammar
            for (Token t : tokens.getTokens()) {
                System.out.println(formatToken(t));
            }

            parser.file_input(); // Test the parser grammar
            System.exit(parser.getNumberOfSyntaxErrors());

        } catch (Exception ex) {
            System.err.println("Error: " + ex.getMessage());
            System.exit(1); // Error occurred, returning non-zero exit code
        }
    }

    private static String formatToken(Token token) {
        String tokenText = replaceSpecialCharacters(token.getText());
        String tokenName = token.getType() == Token.EOF ? "EOF" : PythonLexer.VOCABULARY.getSymbolicName(token.getType());
        String channelText = token.getChannel() == Token.DEFAULT_CHANNEL ?
                             "" :
                             "channel=" + PythonLexer.channelNames[token.getChannel()] + ",";

        // Modified format: [@TokenIndex,StartIndex:StopIndex='Text',<TokenName>,channel=ChannelName,Line:Column]
        return String.format("[@%d,%d:%d='%s',<%s>,%s%d:%d]",
                             token.getTokenIndex(), token.getStartIndex(), token.getStopIndex(),
                             tokenText, tokenName, channelText, token.getLine(), token.getCharPositionInLine());
    }

    private static String replaceSpecialCharacters(String text) {
        return text.replace("\n", "\\n")
               .replace("\r", "\\r")
               .replace("\t", "\\t")
               .replace("\f", "\\f");

    }
}

File: port_CSharp/grun4py.cs
// ******* GRUN (Grammar Unit Test) for Python *******

using System.Text;
using Antlr4.Runtime;

namespace grun4py
{
    internal static class Program
    {
        public static int Main(string[] args)
        {
            if (args.Length < 1)
            {
                Console.Error.WriteLine("Error: Please provide an input file path");
                return 1;
            }

            try
            {
                var filePath = args[0];
                var input = CharStreams.fromPath(filePath, Encoding.GetEncoding("utf-8"));
                var lexer = new PythonLexer(input);
                var tokens = new CommonTokenStream((ITokenSource)lexer);
                var parser = new PythonParser(tokens);

                tokens.Fill(); // Test the lexer grammar
                foreach (IToken t in tokens.GetTokens())
                {
                    Console.WriteLine(FormatToken(t));
                }

                parser.file_input(); // Test the parser grammar
                return parser.NumberOfSyntaxErrors;

            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                return 1; // Error occurred, returning non-zero exit code
            }
        }

        private static string FormatToken(IToken token)
        {
            string tokenText = ReplaceSpecialCharacters(token.Text);
            string tokenName = token.Type == TokenConstants.EOF ? "EOF" : PythonLexer.DefaultVocabulary.GetDisplayName(token.Type);
            string channelText = token.Channel == TokenConstants.DefaultChannel ?
                                 "" :
                                 $"channel={PythonLexer.channelNames[token.Channel]},";

            // Modified format: [@TokenIndex,StartIndex:StopIndex='Text',<TokenName>,channel=ChannelName,Line:Column]
            return $"[@{token.TokenIndex},{token.StartIndex}:{token.StopIndex}='{tokenText}',<{tokenName}>,{channelText}{token.Line}:{token.Column}]";
        }

        private static string ReplaceSpecialCharacters(string text)
        {
            return text.Replace("\n", @"\n")
                       .Replace("\r", @"\r")
                       .Replace("\t", @"\t")
                       .Replace("\f", @"\f");

        }
    }
}
